doc/direction.md

```diff
@@ -0,0 +1,87 @@
+# Minetest Direction Document
+
+---------------------------
+
+## 0. Preface
+
+### 0.1. Introduction
+
+This document attempts to set out Minetest's philosophy, aims, and objectives.
+
+Please note that Minetest is developed by a community, and not everyone in the community
+holds the same views as to what Minetest should be.
+
+### 0.2. Definitions
+
+* `Minetest` is the game engine.
+* `Minetest Game` is the default game which comes with the engine, and is not covered by this document.
+* The `Minetest Project` refers to the combination of the above, the development team, and supporting tools.
+
+### 0.3. Other documents
+
+In order to understand this document, it's important to get context.
+
+celeron55 has published his opinions on the development of Minetest in these places:
+
+  * What is Minetest? - http://c55.me/blog/?p=1491
+  * Roadmap - https://forum.minetest.net/viewtopic.php?t=9177
+  * A clear mission statement for Minetest is missing - https://github.com/minetest/minetest/issues/3476#issuecomment-167399287
+
+Also see the roadmaps and manifestos of core developers on the forums: https://forum.minetest.net/viewforum.php?f=7
+
+---------------------------
+
+## 1. Mission Statement
+
+Minetest provides both a game engine, which allows the creation of voxel games using Lua, and
+a content distribution platform. Players can connect to servers which provide their own gameplay,
+and easily install content through the Minetest client. The goal is to provide an easy to learn and use,
+but still powerful system which allows to easily and rapidly create, extend and modify voxel games.
+
+## 2. Aims
+
+1. **Stability and performance**,
+		especially on older devices.
+2. **Keep the engine as universal as possible**.
+		It's worth comparing everything that goes into the engine for how well
+		it suits a gravitation-less space game without ground surface.
+		If everything is either disableable or suitable for such, then it gives
+		much more freedom for people to experiment and create interesting things,
+		even in ground-based worlds.
+3. **Keep the engine lightweight**.
+		Support a large amount of gameplay by pushing code and functionality out
+		to Lua, rather than having a complex core. Keeping the core as small as
+		possible makes it easier to test and more flexible.
```

Comment:
I'd like to have some balance mentioned here
since performance also matters, moving some features into the core <b>is</b> a good idea (even if lua could technically do it on its own)

Reply:
i don't agree with keeping engine lightweight, it's a wrong direction regarding the performance. I'd rather want to see generic more than lightweight. By generic i mean no explicit object, like firespread, chests or something like this.
But higher level Lua calls are always needed for the performance, and we may need also some frameworks to help minetest to become a game. We are not a game engine, we are mainly tied to voxel games, and we should offer features needed to do a finished an performant game for end users.

Reply:
The core complexity is currently due to a big code layering due to historic and non refactored code from the 0.3 era when minetest moved from a single game to a multi player game. We worked hard to improve it, but we have some layers which are currently crazy to maintain. 
For me, mapgen is one of the good candidates for a refactor (the mapblock generation process), but also the network layer, and the ServerEnvironment.